### DISCUSSION TITLE
Hand.IsGripped/IsPinched and other values when hand is not in view

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#366] [Q&A]
Just noticed on the XR-3 that when my right (or left) hand is not in view (and thus not tracked) both `Input.Hand(Handed.Right).IsGripped` and `Input.Hand(Handed.Right).IsPinched` return `true`. Is the handling of those values when no hand is detected done at the OpenXR layer, or in StereoKit? 

Edit: `IsTracked` is unfortunately also always `true` so can't be used to filter the above situation

**Comment by paulmelis** [comment]
As I noticed before the hand palm position returned when hand tracking is lost seems to be very close to the head position. So I figured I can take advantage of that to do a poor man's "is hand being tracked?" check by looking at the value of `(hand.palm.position - Input.Head.position).MagnitudeSq`. If that's too small then hand tracking is lost.

However, I don't immediately see how I can integrate that test with the built-in `Handle` support, in order to improve the grabbing behaviour when tracking is lost.

**Reply by paulmelis** [reply]
FYI, I've reported the jumps-to-head-when-tracking-lost issue with Varjo support. I'll update here if I get some news on that.

**Comment by paulmelis** [comment]
I wanted to look closer at some of the values coming from the OpenXR layer and was searching for simple standalone example I could hack. And after some googling I found.... https://github.com/maluoi/OpenXRSamples :) Haven't gotten it to work yet though ~~(no console output, seems to exit quickly without any clear reason)~~, I see no rendered cubes, even though controller tracking works including placement of the cubes, so I might open an issue in that repo ;-).

**Reply by maluoi** [reply]
I haven't updated that repo in a while! It _should_ work, but it may have drifted over the years :) I'll have to pop that back open and see if anything obvious has broken!

**Reply by paulmelis** [reply]
With the bit of testing I did yesterday most geometry values, render-related matrices and tracking info seemed to be ok. But for some reason I did not see anything actually getting rendered.

**Reply by paulmelis** [reply]
By the way, your code seems to be only working example of OpenXR usage I could find, so please keep it around! There's too few of those available.

**Reply by maluoi** [reply]
I did just double check it on my computer, and it appears to work as intended on my machine. I guess one thing to note is that there isn't much that gets rendered except for the controllers, and the cubes placed by the controllers! If you're using hand tracking instead of controllers, there's a very good chance you won't see anything at all!

You can also try this in the Mixed Reality Portal, with their headset simulator. Similar problem there, you do need to move the head around a bit in order to see the simulated controllers as well.

**Reply by paulmelis** [reply]
Yes, I noticed it by default only draws two cubes. I turned on the controllers of the XR-3 and got sensible coordinates out of that, so would expect cubes to be rendered at those locations (I checked 360 around in the headset ;-)). Plus, it seems true OpenXR hand tracking isn't enabled as the extension isn't included, right? I did start to hack a bit on that, but it's OpenXR stuff I don't really know.

**Reply by maluoi** [reply]
Yeah, at the time that tutorial was written, there was no such thing as a hand tracking EXT :D And even when it did first come along, HoloLens actually provides a `khr/simple_controller` based on hand tracking data, so this code was sufficient even in that case! I had kinda hoped to continue the "single file" OpenXR series for individual items like the hand tracking EXT, but I started StereoKit soon after that instead.

I do wonder if this is some sort of compositing issue related to Varjo's setup, or if it's a GPU thing instead? Trying it in the WMR Portal would help rule out GPU related issues.

**Reply by paulmelis** [reply]
What exactly should I test in the WMR Portal? I've never used it, nor is the Varjo a Windows Mixed Reality headset, right? I'm just slightly confused what to do exactly.

**Comment by maluoi** [comment]
This is definitely an interesting question, and I suspect there's a lot of runtime-specific behavior that may be mixing itself in here! It's been a while since I was doing this code, but I've also heard some questions from other people about hand tracking related to tracked/untracked state.

I think the first thing that's complicated is that either the OpenXR spec, or runtimes don't do a great job about reporting tracked/untracked hands. I don't think I ever found a nice reliable flag to say whether or not a hand is tracked. And, looking at the code right now, there might be some fiddly pathways through that logic that I should review? Might be fine.

Hand IsTracked is currently based on the locatability of the first hand joint's position! If it's available, the hand is tracked, and if it's not, the hand is not tracked. If a runtime is providing different locatability flags for different joints, this could be problematic! That doesn't seem to be something with a guarantee, but I don't know if it's realistic to expect varying locatability either. In the absence of a better heuristic, it's what I'm using now. (There is an isActive flag, but I remember this as not being a reliable indicator)

The way it _should_ work, is hand data/state should freeze when it becomes untracked. When it becomes tracked again, it will be at its new location. I could imagine a case in which a runtime's hand tracking goes to garbage a frame or two before reporting untracked in some cases, which would be pretty unfortunate. I'd be curious if that's something that's happening, and while I'd be tempted to classify that as a runtime issue, it may be something that happens occasionally regardless?

Anyway, those are some thoughts. I have some outstanding investigations in this area I need to do for someone else as well, so I may have either improvements or more reasons for dismay coming soon :D
